src/replication.c

```diff
@@ -831,34 +1127,93 @@ ask_replica_status_all(spec_t *spec)
 	MTX_UNLOCK(&spec->rq_mtx);
 }
 
-/*
- * This function processes the status response received from replica
- * and update spec's healthy/degrade count
- */
+static void
+handle_prepare_for_rebuild_resp(spec_t *spec, zvol_io_hdr_t *hdr,
+    mgmt_ack_t *ack_data, mgmt_cmd_t *mgmt_cmd)
+{
+
+	int ret = 0;
+	rcommon_mgmt_cmd_t *rcomm_mgmt = mgmt_cmd->rcomm_mgmt;
+	mgmt_ack_t *buf = (mgmt_ack_t *)rcomm_mgmt->buf;
+	
+	if (hdr->status != ZVOL_OP_STATUS_OK) {
+		rcomm_mgmt->cmds_failed++;
+	} else {
+		memcpy(&buf[rcomm_mgmt->cmds_succeeded++], ack_data,
+		    sizeof (mgmt_ack_t));
+	}
+
+	if (rcomm_mgmt->cmds_sent == rcomm_mgmt->cmds_succeeded) {
+		ret = start_rebuild(buf, spec->target_replica, rcomm_mgmt->buf_size);
+		if (ret == 0) {
+			rcomm_mgmt->buf = NULL;
+			REPLICA_LOG("Rebuild triggered on Replica ip:%s port:%d"
+			    " state:%d\n", spec->target_replica->ip,
+			    spec->target_replica->port, spec->target_replica->state);
+		} else {
+			MTX_LOCK(&spec->rq_mtx);
+			spec->target_replica = NULL;
+			spec->rebuild_in_progress = false;
+			MTX_UNLOCK(&spec->rq_mtx);
+			REPLICA_LOG("Unable to trigger rebuild on Replica ip:"
+			    "%s port:%d state:%d\n", spec->target_replica->ip,
+			    spec->target_replica->port, spec->target_replica->state);
+		}	
+		free_rcommon_mgmt_cmd(rcomm_mgmt);
+	} else if (rcomm_mgmt->cmds_sent ==
+	    (rcomm_mgmt->cmds_failed + rcomm_mgmt->cmds_succeeded)) {
+		free_rcommon_mgmt_cmd(rcomm_mgmt);
+		spec->target_replica = NULL;
+		spec->rebuild_in_progress = false;
+	}
+}
+
 static int
 update_replica_status(spec_t *spec, replica_t *replica)
 {
 	zrepl_status_ack_t *repl_status;
-	replica_state_t last_status;
+	replica_state_t last_state;
 
 	repl_status = (zrepl_status_ack_t *)replica->mgmt_io_resp_data;
 
+	REPLICA_ERRLOG("Replica ip:%s port:%d state:%d rebuild status:%d\n",
+	    replica->ip, replica->port, repl_status->state,
+	    repl_status->rebuild_status);
+
 	MTX_LOCK(&spec->rq_mtx);
 	MTX_LOCK(&replica->r_mtx);
-	last_status = replica->state;
+
+	last_state = replica->state;
 	replica->state = (replica_state_t) repl_status->state;
 	MTX_UNLOCK(&replica->r_mtx);
 
-	if(last_status != repl_status->state) {
+	if (last_state != repl_status->state) {
 		if (repl_status->state == ZVOL_STATUS_DEGRADED) {
 			spec->degraded_rcount++;
 			spec->healthy_rcount--;
 		} else if (repl_status->state == ZVOL_STATUS_HEALTHY) {
+			/* master_replica became healthy*/
 			spec->degraded_rcount--;
 			spec->healthy_rcount++;
+			spec->target_replica = NULL;
+			spec->rebuild_in_progress = false;
+			REPLICA_ERRLOG("Replica:%s port:%d marked healthy,"
+		    	    " seting master_replica to NULL\n",
+		    	    replica->ip, replica->port);
 		}
-		update_volstate(spec);
```

Comment:
we would be needing this

Reply:
I think update_volstate should be used only at two places ....
> during replica registration
> during replica disconnection
These are the events which can increase/decrease degraded_counts, so needed there only. Thats where we set spec->ready to true.

Reply:
oops.. I forgot again. you are right sir.